File: languages/python/minimize-boolean-function.py
#!/usr/bin/python3
import math

def minimize_boolean_function(truth_table):
    """
    Minimize a Boolean function of N variables given a truth table with 0s, 1s, and Xs.
    Input: truth_table - list of 2^N values (0, 1, or 'X') for minterms 0 to 2^N-1.
    Output: Simplified Boolean expression as a string.
    """
    # Step 1: Determine number of variables
    n_vars = int(math.log2(len(truth_table)))
    if len(truth_table) != 2 ** n_vars:
        raise ValueError("Truth table length must be a power of 2")

    # Step 2: Extract minterms (1s) and don't cares (Xs)
    minterms = [i for i, val in enumerate(truth_table) if val == 1]
    dont_cares = [i for i, val in enumerate(truth_table) if val == 'X']
    if not minterms:
        return "0"  # No 1s in truth table

    # Step 3: Generate implicants (minterms + don't cares)
    #print("Step 3: minterms =", minterms, ", dont_cares =", dont_cares)
    implicants = minterms + dont_cares
    if not implicants:
        return "0"

    # Step 4: Group implicants by number of 1s in binary representation
    def count_ones(n):
        if isinstance(n, str):
            return n.count('1')  # For combined terms like '10-1'
        return bin(n).count('1')  # For integer minterms

    groups = {i: [] for i in range(n_vars + 1)}  # 0 to N ones
    for m in implicants:
        groups[count_ones(m)].append((m, bin(m)[2:].zfill(n_vars)))

    # Step 5: Quine-McCluskey: Combine implicants differing by one bit
    def differ_by_one(s1, s2):
        return sum(a != b for a, b in zip(s1, s2)) == 1

    def combine_terms(m1, m2, s1, s2):
        result = ''.join('-' if a != b else a for a, b in zip(s1, s2))
        return result

    prime_implicants = set()
    used = set()
    while True:
        new_groups = {i: [] for i in range(n_vars + 1)}
        combined = set()
        for i in range(n_vars):
            for m1, s1 in groups[i]:
                for m2, s2 in groups[i + 1]:
                    if differ_by_one(s1, s2):
                        new_term = combine_terms(m1, m2, s1, s2)
                        new_groups[count_ones(new_term)].append((None, new_term))
                        combined.add((m1, s1))
                        combined.add((m2, s2))
                        if m1 is not None:
                            used.add(m1)
                        if m2 is not None:
                            used.add(m2)
        # Add uncombined terms as prime implicants
        for i in range(n_vars + 1):
            for m, s in groups[i]:
                if (m, s) not in combined:
                    prime_implicants.add(s)
        # Update groups
        groups = new_groups
        if not any(groups[i] for i in range(n_vars + 1)):
            break
    # Convert to list for repeatability
    prime_implicants = sorted(prime_implicants)

    # Step 6: Select essential prime implicants
    #print("Step 6: prime_implicants =", prime_implicants)
    def covers_implicant(term, minterm):
        term_bits = term
        minterm_bits = bin(minterm)[2:].zfill(n_vars)
        return all(t == '-' or t == m for t, m in zip(term_bits, minterm_bits))

    essential = set()
    covered = set()
    for m in minterms:
        if m in dont_cares:
            continue
        covering = [pi for pi in prime_implicants if covers_implicant(pi, m)]
        if len(covering) == 1:
            essential.add(covering[0])
            covered.update([m for m in minterms if covers_implicant(covering[0], m)])
    # Add additional implicants if needed
    for m in minterms:
        if m not in covered and m not in dont_cares:
            for pi in prime_implicants:
                if pi not in essential and covers_implicant(pi, m):
                    essential.add(pi)
                    covered.update([m for m in minterms if covers_implicant(pi, m)])
                    break
    # Convert to list for repeatability
    essential = sorted(essential, reverse=True)

    # Step 7: Convert prime implicants to Boolean expression
    #print("Step 7: essential =", essential)
    def term_to_expression(term):
        vars = [chr(65 + i) for i in range(n_vars)]  # A, B, C, ...
        result = []
        for v, t in zip(vars, term):
            if t == '1':
                result.append(v)
            elif t == '0':
                result.append(f"~{v}")
        return ''.join(result) if result else "1"

    if not essential:
        return "0"
    expression = ' + '.join(term_to_expression(pi) for pi in essential)
    return expression if expression else "1"

# Example usage
if __name__ == "__main__":
    #
    # Example: 2-variable truth table (4 entries)
    #
    input = [0, 1, 0, 1]
    result = minimize_boolean_function(input)
    print("2-variable input:", input)
    print("          result:", result)
    assert result == "B"

    #
    # Example: 3-variable truth table (8 entries)
    #
    input = [0, 1, 'X', 0, 1, 0, 'X', 1]
    result = minimize_boolean_function(input)
    print("3-variable input:", input)
    print("          result:", result)
    assert result == "AB + A~C + ~A~BC"

    input = [0, 1, 0, 1, 'X', 1, 'X', 0]
    result = minimize_boolean_function(input)
    print("3-variable input:", input)
    print("          result:", result)
    assert result == "~AC + ~BC"

    #
    # Example: 4-variable truth table (16 entries)
    #
    input = [0, 1, 'X', 0, 1, 0, 0, 1, 'X', 0, 0, 1, 0, 1, 'X', 1]
    result = minimize_boolean_function(input)
    print("4-variable input:", input)
    print("          result:", result)
    assert result == "ABD + ACD + ~AB~C~D + ~A~B~CD + BCD"

    # Truth table for f(A,B,C,D) = Σ(0,5,10,15) + d(2,7,12)
    #        0  1   2   3  4  5  6   7   8  9 10 11  12  13 14 15
    input = [1, 0, 'X', 0, 0, 1, 0, 'X', 0, 0, 1, 0, 'X', 0, 0, 1]
    result = minimize_boolean_function(input)
    print("4-variable input:", input)
    print("          result:", result)
    assert result == "~ABD + ~A~B~D + BCD + ~BC~D"

    #
    # Example: 5-variable truth table (32 entries)
    #
    # Truth table for f(A,B,C,D,E) = Σ(3,9,17,25,31) + d(0,10,20,30)
    #         0   1  2  3  4  5  6  7  8  9  10  11 12 13 14 15 16 17 18 19  20  21 22 23 24 25 26 27 28 29  30  31
    input = ['X', 0, 0, 1, 0, 0, 0, 0, 0, 1, 'X', 0, 0, 0, 0, 0, 0, 1, 0, 0, 'X', 0, 0, 0, 0, 1, 0, 0, 0, 0, 'X', 1]
    result = minimize_boolean_function(input)
    print("5-variable input:", input)
    print("          result:", result)
    assert result == "ABCD + A~C~DE + ~A~B~CDE + B~C~DE"

    #
    # Example: 8-variable truth table (256 entries)
    #
    input = [
        'X' if i in [0, 1, 2, 3, 4] else
        1 if i in [10, 100, 255] else
        0
        for i in range(256)
    ]
    result = minimize_boolean_function(input)
    print("8-variable input:", input)
    print("          result:", result)
    assert result == "ABCDEFGH + ~ABC~D~EF~G~H + ~A~B~C~D~FG~H"

File: languages/rust/minimize-boolean-function.rs
use std::collections::{HashSet, HashMap};
use std::char;

#[derive(Clone, PartialEq, Eq, Hash)]
enum TruthValue {
    Zero,
    One,
    DontCare,
}

fn parse_truth_table(input: &str) -> Result<Vec<TruthValue>, String> {
    input.chars().map(|c| match c {
        '0' => Ok(TruthValue::Zero),
        '1' => Ok(TruthValue::One),
        'X' | 'x' => Ok(TruthValue::DontCare),
        _ => Err(format!("Invalid character in truth table: {}", c)),
    }).collect()
}

fn minimize_boolean_function(truth_table_str: &str) -> Result<String, String> {
    // Step 1: Parse truth table string
    let truth_table = parse_truth_table(truth_table_str)?;

    // Step 2: Determine number of variables
    let n_vars = (truth_table.len() as f64).log2() as usize;
    if truth_table.len() != 1 << n_vars {
        return Err("Truth table length must be a power of 2".to_string());
    }

    // Step 3: Extract minterms (1s) and don't cares (Xs)
    let minterms: Vec<usize> = truth_table.iter().enumerate()
        .filter(|(_, v)| **v == TruthValue::One)
        .map(|(i, _)| i)
        .collect();
    let dont_cares: Vec<usize> = truth_table.iter().enumerate()
        .filter(|(_, v)| **v == TruthValue::DontCare)
        .map(|(i, _)| i)
        .collect();
    if minterms.is_empty() {
        return Ok("0".to_string());
    }

    // Step 4: Generate implicants (minterms + don't cares)
    let mut implicants = minterms.clone();
    implicants.extend(dont_cares.iter());
    if implicants.is_empty() {
        return Ok("0".to_string());
    }

    // Step 5: Group implicants by number of 1s
    fn count_ones(n: &str) -> usize {
        n.chars().filter(|&c| c == '1').count()
    }

    fn count_ones_int(n: usize) -> usize {
        n.count_ones() as usize
    }

    let mut groups: HashMap<usize, Vec<(Option<usize>, String)>> = HashMap::new();
    for i in 0..=n_vars {
        groups.insert(i, vec![]);
    }
    for &m in &implicants {
        let binary = format!("{:0width$b}", m, width=n_vars);
        groups.get_mut(&count_ones_int(m)).unwrap().push((Some(m), binary));
    }

    // Step 6: Quine-McCluskey: Combine implicants differing by one bit
    fn differ_by_one(s1: &str, s2: &str) -> bool {
        s1.chars().zip(s2.chars()).filter(|&(a, b)| a != b).count() == 1
    }

    fn combine_terms(s1: &str, s2: &str) -> String {
        s1.chars().zip(s2.chars())
            .map(|(a, b)| if a != b { '-' } else { a })
            .collect()
    }

    let mut prime_implicants: HashSet<String> = HashSet::new();
    let mut used: HashSet<usize> = HashSet::new();
    loop {
        let mut new_groups: HashMap<usize, Vec<(Option<usize>, String)>> = HashMap::new();
        for i in 0..=n_vars {
            new_groups.insert(i, vec![]);
        }
        let mut combined: HashSet<(Option<usize>, String)> = HashSet::new();
        for i in 0..n_vars {
            for (m1, s1) in groups.get(&i).unwrap_or(&vec![]) {
                for (m2, s2) in groups.get(&(i + 1)).unwrap_or(&vec![]) {
                    if differ_by_one(s1, s2) {
                        let new_term = combine_terms(s1, s2);
                        new_groups.get_mut(&count_ones(&new_term)).unwrap()
                            .push((None, new_term.clone()));
                        combined.insert((m1.clone(), s1.clone()));
                        combined.insert((m2.clone(), s2.clone()));
                        if let Some(m1_val) = m1 {
                            used.insert(*m1_val);
                        }
                        if let Some(m2_val) = m2 {
                            used.insert(*m2_val);
                        }
                    }
                }
            }
        }
        // Add uncombined terms as prime implicants
        for i in 0..=n_vars {
            for (m, s) in groups.get(&i).unwrap_or(&vec![]) {
                if !combined.contains(&(*m, s.clone())) {
                    prime_implicants.insert(s.clone());
                }
            }
        }
        // Update groups
        groups = new_groups;
        if groups.values().all(|v| v.is_empty()) {
            break;
        }
    }

    // Step 7: Select essential prime implicants
    fn covers_implicant(term: &str, minterm: usize, n_vars: usize) -> bool {
        let minterm_bits = format!("{:0width$b}", minterm, width=n_vars);
        term.chars().zip(minterm_bits.chars())
            .all(|(t, m)| t == '-' || t == m)
    }

    let mut essential: Vec<String> = vec![];
    let mut covered: HashSet<usize> = HashSet::new();
    for &m in &minterms {
        if dont_cares.contains(&m) {
            continue;
        }
        let covering: Vec<String> = prime_implicants.iter()
            .filter(|pi| covers_implicant(pi, m, n_vars))
            .cloned()
            .collect();
        if covering.len() == 1 {
            let pi = covering[0].clone();
            if !essential.contains(&pi) {
                essential.push(pi.clone());
            }
            for m2 in minterms.iter().filter(|m2| covers_implicant(&pi, **m2, n_vars)) {
                covered.insert(*m2);
            }
        }
    }
    // Add additional implicants if needed
    for &m in &minterms {
        if !covered.contains(&m) && !dont_cares.contains(&m) {
            for pi in &prime_implicants {
                if !essential.contains(pi) && covers_implicant(pi, m, n_vars) {
                    essential.push(pi.clone());
                    for m2 in minterms.iter().filter(|m2| covers_implicant(pi, **m2, n_vars)) {
                        covered.insert(*m2);
                    }
                    break;
                }
            }
        }
    }

    // Step 8: Convert prime implicants to Boolean expression
    fn term_to_expression(term: &str, n_vars: usize) -> String {
        let vars: Vec<char> = (0..n_vars).map(|i| char::from_u32('A' as u32 + i as u32).unwrap()).collect();
        let mut result: Vec<String> = vec![];
        for (v, t) in vars.iter().zip(term.chars()) {
            match t {
                '1' => result.push(v.to_string()),
                '0' => result.push(format!("~{}", v)),
                _ => {},
            }
        }
        if result.is_empty() { "1".to_string() } else { result.join("") }
    }

    if essential.is_empty() {
        return Ok("0".to_string());
    }
    let expression = essential.iter()
        .map(|pi| term_to_expression(pi, n_vars))
        .collect::<Vec<String>>()
        .join(" + ");
    Ok(if expression.is_empty() { "1".to_string() } else { expression })
}

// Helper function to generate 8-variable truth table
fn generate_8var_truth_table() -> String {
    let mut table = vec!['0'; 256];
    // Pattern 1: A=1, B=0, C=1, D=1 (E,F,G,H any) -> 1 or X
    for efgh in 0..16 {
        let minterm = (1 << 7) | (0 << 6) | (1 << 5) | (1 << 4) | efgh; // 1011xxxx
        table[minterm] = if efgh % 2 == 0 { '1' } else { 'X' };
    }
    // Pattern 2: E=1, F=0, G=1 (A,B,C,D,H any) -> 1 or X
    for abcdh in 0..32 {
        let minterm = (abcdh >> 4) << 7 | (1 << 3) | (0 << 2) | (1 << 1) | (abcdh & 1); // xxxx101x
        table[minterm] = if abcdh % 3 == 0 { '1' } else { 'X' };
    }
    // Pattern 3: A=0, B=1, H=1 (C,D,E,F,G any) -> 1 or X
    for cdefg in 0..32 {
        let minterm = (0 << 7) | (1 << 6) | (cdefg << 1) | 1; // 01xxxxx1
        table[minterm] = if cdefg % 2 == 0 { '1' } else { 'X' };
    }
    table.into_iter().collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_3_variables() {
        let truth_table = "01X010X1";
        let result = minimize_boolean_function(truth_table).unwrap();
        println!("3-variable result: {}", result);
    }

    #[test]
    fn test_4_variables() {
        let truth_table = "01X01001X00101X1";
        let result = minimize_boolean_function(truth_table).unwrap();
        println!("4-variable result: {}", result);
    }

    #[test]
    fn test_8_variables() {
        let truth_table = generate_8var_truth_table();
        let result = minimize_boolean_function(&truth_table).unwrap();
        println!("8-variable result: {}", result);
    }
}

fn main() {
    // Example usage for 8-variable function
    let truth_table_8 = generate_8var_truth_table();
    println!("Truth table: {}", truth_table_8);
    match minimize_boolean_function(&truth_table_8) {
        Ok(result) => println!("8-variable result: {}", result),
        Err(e) => println!("Error: {}", e),
    }
}
